storage/replica.go

```diff
@@ -1660,3 +1677,49 @@ func (r *Replica) maybeAddToSplitQueue() {
 		r.store.splitQueue.MaybeAdd(r, r.store.Clock().Now())
 	}
 }
+
+// processReplicaCmd get a raft command from command queue and
+// call processRaftCommand to execute in a loop until there is
+// not any command in the queue for 5 seconds.
+func (r *Replica) processReplicaCmd() {
+	for {
+		select {
+		case cmd := <-r.applyQueueChan:
+			err := r.processRaftCommand(cmd.idKey, cmd.index, cmd.raftCmd)
+			if cmd.callback != nil {
+				cmd.callback(err)
+			}
+
+		case <-time.After(5 * time.Second):
+			r.stopFlagLock.Lock()
+			if len(r.applyQueueChan) == 0 {
```

Comment:
Checking the `len` of a channel is racy and I don't think you need it here. You know that the channel is empty because you hit the `time.After` instead of reading from the channel. Just set `stopFlag` and return. If something is concurrently added to the channel it will see `stopFlag` and recreate the goroutine.